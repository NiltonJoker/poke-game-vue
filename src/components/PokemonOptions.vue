<template>
  <div class="options-container">
    <ul class="fade-in" v-if="pokemons.length !== 0">
      <li
        v-for="pokemon in pokemons"
        :key="pokemon.id"
        @click="$emit('selection', pokemon.id)"
      >
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  props: {
    pokemons: {
      type: Array,
      required: true,
    },
  },

  methods: {},
};
</script>

<style scoped>
ul {
  padding: 0;
  list-style-type: none;
  display: flex;
  flex-direction: column;
  gap: 10px;
}
li {
  background-color: white;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  padding: 8px 16px;
  width: 250px;
  text-transform: capitalize;
}

li:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
  display: flex;
  justify-content: center;
}
</style>