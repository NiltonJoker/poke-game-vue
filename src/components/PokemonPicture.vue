<template>
  <div class="pokemon-container">
    <img
      :src="imgSrc"
      class="hidden-pokemon"
      alt="pokemon"
    />
    <img
      v-if="showPokemon"
      :src="imgSrc"
      class="fade-in"
      alt="pokemon"
    />
  </div>
</template>

<script>
export default {

  props: {
    pokemonId: {
      type: Number,
      required: true,
    },
    showPokemon: {
      type: Boolean,
      required: true,
      default: false
    }
  },

  computed: {
    imgSrc() {
      return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonId}.svg`
    }
  }

};
</script>

<style scoped>
.pokemon-container {
  height: 200px;
  display: flex;
  justify-content: center;
}


 @media (max-width: 768px){
  .pokemon-container {
    height: 100px;
  }
  .pokemon-container img {
    max-height: 100px;
  }
}

.hidden-pokemon {
  filter: brightness(0);
}

img {
  display: block;
  height: 200px;
  position: absolute;
  width: 200px;
  user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-drag: none;
  -webkit-user-select: none;
}
</style>