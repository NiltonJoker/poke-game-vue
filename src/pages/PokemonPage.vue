<template>
  <!-- <h1 v-if="!pokemon">Espere por favor...</h1> -->
  <PokemonLoading v-if="!pokemon" />

  <div class="pokemon-page fade-in" v-else>
    <PokemonHeader :aciertos="aciertos" :desaciertos="desaciertos" />

    <h2>¿Quién es ese Pokemón?</h2>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />

    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />

    <div v-if="showAnswer" class="fade-in">
      <h2 class="message-answer">{{ message }}</h2>
      <button class="btn-newgame" @click="newGame">Nuevo Juego</button>
    </div>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";
import PokemonLoading from "@/components/PokemonLoading.vue";
import PokemonHeader from "@/components/PokemonHeader.vue";

export default {
  components: { PokemonPicture, PokemonOptions, PokemonLoading, PokemonHeader },

  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
      aciertos: 0,
      desaciertos: 0,
    };
  },
  methods: {
    async mixedPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const rndInt = Math.floor(Math.random() * 4);

      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(pokemonSelected) {
      if (this.showAnswer) return;

      this.showAnswer = true;
      this.showPokemon = true;

      if (pokemonSelected === this.pokemon.id) {
        this.message = `Correcto, ${this.pokemon.name}`;
        this.aciertos++;
      } else {
        this.message = `Oops, era ${this.pokemon.name}`;
        this.desaciertos++;
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemon = null;
      (this.pokemonArr = []), this.mixedPokemonArray();
    },
  },
  mounted() {
    this.mixedPokemonArray();
  },
};
</script>


<style scoped>
.title-page {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 300px;
  height: 100px;
  object-fit: contain;
}

.message-answer {
  text-transform: capitalize;
}
.pokemon-page {
  width: 100%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  /* justify-content: center; */
  align-items: center;
  gap: 10px;
  margin-top: 20px;
}
.btn-newgame {
  border: none;
  padding: 8px 16px;
  border-radius: 5px;

  font-size: 15px;
  transition: all 0.3s ease-in-out;
  background-color: #e3350d;
  color: white;
  cursor: pointer;
}
</style>