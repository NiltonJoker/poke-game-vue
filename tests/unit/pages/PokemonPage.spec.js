import PokemonPage from "@/pages/PokemonPage";
import PokemonPicture from "@/components/PokemonPicture";
import PokemonOptions from "@/components/PokemonOptions";

import { shallowMount, mount } from "@vue/test-utils";
import { pokemons } from "../mocks/pokemons.mock";

describe("PokemonPage Component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test("debe de hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de llamar el mixedPokemonArr al montar", () => {
    const mixedPokemonArrSpy = jest.spyOn(
      PokemonPage.methods,
      "mixedPokemonArray"
    );

    shallowMount(PokemonPage);

    expect(mixedPokemonArrSpy).toHaveBeenCalled();
  });

  test("Debe de hacer match con el snapshot cuando cargan los pokemons", () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          aciertos: 0,
          desaciertos: 0,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Debe de mostrar los componentes de PokemonPicture y PokemonOptions", () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          aciertos: 0,
          desaciertos: 0,
        };
      },
    });

    const picture = wrapper.findComponent(PokemonPicture);
    const options = wrapper.findComponent(PokemonOptions);

    expect(picture.exists()).toBe(true);
    expect(options.exists()).toBe(true);

    expect(picture.props("pokemonId")).toBe(1);
    expect(Boolean(options.props("pokemons"))).toBe(true);
  });

  test('Pruebas con checkAnswer', async () => {
    
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          aciertos: 0,
          desaciertos: 0,
        };
      },
    });

    await wrapper.vm.checkAnswer(1)

    expect(wrapper.find('h2.message-answer').exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.find('h2.message-answer').text()).toBe(`Correcto, ${pokemons[0].name}`)

  })

  test('Pruebas con checkAnswer sino es correcto', async () => {
    
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          aciertos: 0,
          desaciertos: 0,
        };
      },
    });

    await wrapper.vm.checkAnswer(2)

    expect(wrapper.find('h2.message-answer').exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`)

  })
  
});
