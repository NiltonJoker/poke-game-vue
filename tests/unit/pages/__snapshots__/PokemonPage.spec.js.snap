// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<!-- <h1 v-if="!pokemon">Espere por favor...</h1> -->
<div class="pokemon-page fade-in"><img class="title-page" src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/98/International_Pok%C3%A9mon_logo.svg/375px-International_Pok%C3%A9mon_logo.svg.png" alt="">
  <div class="pokemon-question">
    <div>
      <p class="text-success">Aciertos: 0</p>
      <p class="text-error">Desaciertos: 0</p>
    </div><img class="pokeball-img" src="https://i.pinimg.com/originals/09/da/92/09da926c2b94d95008a9e3b2f60bfdd3.png" alt="">
  </div>
  <h2>¿Quién es ese Pokemón?</h2>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" class="hidden-pokemon" alt="pokemon">
    <!--v-if-->
  </div>
  <div class="options-container">
    <ul class="fade-in">
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `
<!-- <h1 v-if="!pokemon">Espere por favor...</h1> -->
<pokemon-loading-stub></pokemon-loading-stub>
`;
